### ISSUE TITLE
M1 Mac support

### ISSUE THREAD

**Issue by limjiyoon** [kahypar/kahypar#98]
Is there any plan for supporting m1 mac?
When I tried to install kahypar in m1 mac, it occurs doesn't support cpu=vortex error.

**Comment by SebastianSchlag** [comment]
In principle, we should be able to support M1 macs as there are few CPU specific things. Could you provide us with a more detailed error message such that we can understand what needs to be changed?

**Comment by limjiyoon** [comment]
I build kahypar according to the https://github.com/kahypar/kahypar#building-kahypar.

First it shows following error when I run `make`
`clang: error: the clang compiler does not support '-march=native'`

When I search for it, there is no clang target for m1. 
When I change the `--march=natvie` to `-mcpu=apple-a14` in CmakeList.txt,
Then it shows following error messages.
```/Library/Developer/CommandLineTools/usr/lib/clang/12.0.5/include/ia32intrin.h:188:10: error: use of undeclared identifier '__builtin_ia32_readeflags_u32'
  return __builtin_ia32_readeflags_u32();
         ^
/Library/Developer/CommandLineTools/usr/lib/clang/12.0.5/include/ia32intrin.h:194:3: error: use of undeclared identifier '__builtin_ia32_writeeflags_u32'
  __builtin_ia32_writeeflags_u32(__f);
  ^
/Library/Developer/CommandLineTools/usr/lib/clang/12.0.5/include/ia32intrin.h:284:10: error: use of undeclared identifier '__builtin_ia32_crc32qi'
  return __builtin_ia32_crc32qi(__C, __D);
         ^
/Library/Developer/CommandLineTools/usr/lib/clang/12.0.5/include/ia32intrin.h:305:10: error: use of undeclared identifier '__builtin_ia32_crc32hi'; did you mean '__builtin_arm_crc32h'?
  return __builtin_ia32_crc32hi(__C, __D);
         ^
/Library/Developer/CommandLineTools/usr/lib/clang/12.0.5/include/ia32intrin.h:305:10: note: '__builtin_arm_crc32h' declared here
/Library/Developer/CommandLineTools/usr/lib/clang/12.0.5/include/ia32intrin.h:326:10: error: use of undeclared identifier '__builtin_ia32_crc32si'
  return __builtin_ia32_crc32si(__C, __D);
         ^
/Library/Developer/CommandLineTools/usr/lib/clang/12.0.5/include/ia32intrin.h:354:10: error: use of undeclared identifier '__builtin_ia32_rdpmc'; did you mean '__builtin_arm_dmb'?
  return __builtin_ia32_rdpmc(__A);
         ^
/Library/Developer/CommandLineTools/usr/lib/clang/12.0.5/include/ia32intrin.h:354:10: note: '__builtin_arm_dmb' declared here
/Library/Developer/CommandLineTools/usr/lib/clang/12.0.5/include/ia32intrin.h:354:10: error: argument to '__builtin_arm_dmb' must be a constant integer
  return __builtin_ia32_rdpmc(__A);
         ^                    ~~~
/Library/Developer/CommandLineTools/usr/lib/clang/12.0.5/include/ia32intrin.h:360:10: error: use of undeclared identifier '__builtin_ia32_rdtscp'; did you mean '__builtin_arm_rsrp'?
  return __builtin_ia32_rdtscp(__A);
         ^
/Library/Developer/CommandLineTools/usr/lib/clang/12.0.5/include/ia32intrin.h:360:10: note: '__builtin_arm_rsrp' declared here
/Library/Developer/CommandLineTools/usr/lib/clang/12.0.5/include/ia32intrin.h:360:32: error: cannot initialize a parameter of type 'const char *' with an lvalue of type 'unsigned int *'
  return __builtin_ia32_rdtscp(__A);
                               ^~~
/Library/Developer/CommandLineTools/usr/lib/clang/12.0.5/include/ia32intrin.h:369:3: error: use of undeclared identifier '__builtin_ia32_wbinvd'
  __builtin_ia32_wbinvd();
  ^
In file included from /Users/jiyoonlim/work/kahypar/kahypar/kahypar/application/kahypar.cc:22:
In file included from /Users/jiyoonlim/work/kahypar/kahypar/kahypar/application/command_line_options.h:37:
In file included from /Users/jiyoonlim/work/kahypar/kahypar/kahypar/kahypar.h:23:
In file included from /Users/jiyoonlim/work/kahypar/kahypar/kahypar/partition/registries/register_coarsening_algorithms.h:24:
In file included from /Users/jiyoonlim/work/kahypar/kahypar/kahypar/partition/coarsening/do_nothing_coarsener.h:25:
In file included from /Users/jiyoonlim/work/kahypar/kahypar/kahypar/definitions.h:27:
In file included from /Users/jiyoonlim/work/kahypar/kahypar/kahypar/datastructure/hypergraph.h:45:
In file included from /Users/jiyoonlim/work/kahypar/kahypar/kahypar/utils/math.h:27:
In file included from /Library/Developer/CommandLineTools/usr/lib/clang/12.0.5/include/x86intrin.h:15:
In file included from /Library/Developer/CommandLineTools/usr/lib/clang/12.0.5/include/immintrin.h:15:
/Library/Developer/CommandLineTools/usr/lib/clang/12.0.5/include/mmintrin.h:33:5: error: use of undeclared identifier '__builtin_ia32_emms'; did you mean '__builtin_isless'?
    __builtin_ia32_emms();
    ^
/Library/Developer/CommandLineTools/SDKs/MacOSX11.3.sdk/usr/include/c++/v1/math.h:649:12: note: '__builtin_isless' declared here
    return isless(__lcpp_x, __lcpp_y);
           ^
/Library/Developer/CommandLineTools/SDKs/MacOSX11.3.sdk/usr/include/math.h:545:22: note: expanded from macro 'isless'
#define isless(x, y) __builtin_isless((x),(y))
                     ^
In file included from /Users/jiyoonlim/work/kahypar/kahypar/kahypar/application/kahypar.cc:22:
In file included from /Users/jiyoonlim/work/kahypar/kahypar/kahypar/application/command_line_options.h:37:
In file included from /Users/jiyoonlim/work/kahypar/kahypar/kahypar/kahypar.h:23:
In file included from /Users/jiyoonlim/work/kahypar/kahypar/kahypar/partition/registries/register_coarsening_algorithms.h:24:
In file included from /Users/jiyoonlim/work/kahypar/kahypar/kahypar/partition/coarsening/do_nothing_coarsener.h:25:
In file included from /Users/jiyoonlim/work/kahypar/kahypar/kahypar/definitions.h:27:
In file included from /Users/jiyoonlim/work/kahypar/kahypar/kahypar/datastructure/hypergraph.h:45:
In file included from /Users/jiyoonlim/work/kahypar/kahypar/kahypar/utils/math.h:27:
In file included from /Library/Developer/CommandLineTools/usr/lib/clang/12.0.5/include/x86intrin.h:15:
In file included from /Library/Developer/CommandLineTools/usr/lib/clang/12.0.5/include/immintrin.h:15:
/Library/Developer/CommandLineTools/usr/lib/clang/12.0.5/include/mmintrin.h:33:25: error: too few arguments to function call, expected 2, have 0
    __builtin_ia32_emms();
                        ^
/Library/Developer/CommandLineTools/usr/lib/clang/12.0.5/include/mmintrin.h:50:19: error: use of undeclared identifier '__builtin_ia32_vec_init_v2si'
    return (__m64)__builtin_ia32_vec_init_v2si(__i, 0);
                  ^
/Library/Developer/CommandLineTools/usr/lib/clang/12.0.5/include/mmintrin.h:67:12: error: use of undeclared identifier '__builtin_ia32_vec_ext_v2si'
    return __builtin_ia32_vec_ext_v2si((__v2si)__m, 0);
           ^
/Library/Developer/CommandLineTools/usr/lib/clang/12.0.5/include/mmintrin.h:129:19: error: use of undeclared identifier '__builtin_ia32_packsswb'
    return (__m64)__builtin_ia32_packsswb((__v4hi)__m1, (__v4hi)__m2);
                  ^
/Library/Developer/CommandLineTools/usr/lib/clang/12.0.5/include/mmintrin.h:159:19: error: use of undeclared identifier '__builtin_ia32_packssdw'
    return (__m64)__builtin_ia32_packssdw((__v2si)__m1, (__v2si)__m2);
                  ^
/Library/Developer/CommandLineTools/usr/lib/clang/12.0.5/include/mmintrin.h:189:19: error: use of undeclared identifier '__builtin_ia32_packuswb'
    return (__m64)__builtin_ia32_packuswb((__v4hi)__m1, (__v4hi)__m2);
                  ^
/Library/Developer/CommandLineTools/usr/lib/clang/12.0.5/include/mmintrin.h:216:19: error: use of undeclared identifier '__builtin_ia32_punpckhbw'
    return (__m64)__builtin_ia32_punpckhbw((__v8qi)__m1, (__v8qi)__m2);
                  ^
/Library/Developer/CommandLineTools/usr/lib/clang/12.0.5/include/mmintrin.h:239:19: error: use of undeclared identifier '__builtin_ia32_punpckhwd'
    return (__m64)__builtin_ia32_punpckhwd((__v4hi)__m1, (__v4hi)__m2);
                  ^
fatal error: too many errors emitted, stopping now [-ferror-limit=]
20 errors generated.
make[2]: *** [kahypar/application/CMakeFiles/KaHyPar.dir/kahypar.cc.o] Error 1
make[1]: *** [kahypar/application/CMakeFiles/KaHyPar.dir/all] Error 2
make: *** [all] Error 2```

**Comment by jeshicawang** [comment]
Same issue for me, would love to be able to compile on Apple M1!

**Comment by SebastianSchlag** [comment]
Hey @jeshicawang and Romain,
I’ll try to find some time to look into this issue. Would you be willing to help testing possible work-arounds for this as this will likely take a couple of iterations to get right? 
I don‘t have an M1 at hand and don’t have any experience with cross compiling.

**Comment by SebastianSchlag** [comment]
So I tried to get this working today.
@jeshicawang @Romain @limjiyoon could someone of you folks try the following:

- Checkout the `m1-support` branch
- go to `kahypar/external_tools/WHFC/util/meta.h` and remove or comment out line 8 (the `#include <emmintrin.h>` statement)
- try to compile

Maybe this is already enough.

@larsgottesbueren it seems like the `<emmintrin.h>` is not used at all in the WHFC version that we use. Could you remove that and update the submodule version accordingly?

**Comment by romainwallon** [comment]
Hi @SebastianSchlag,

I have just tested the steps described above, and I managed to compile KaHyPar on my M1 Mac.
Thanks for this workaround!

I had to use the `-DKAHYPAR_USE_MINIMAL_BOOST=ON` option to make it work, but maybe this is because my boost installation is not fully functional.

**Comment by larsgottesbueren** [comment]
I've updated the submodule for the m1-support branch.

**Comment by SebastianSchlag** [comment]
Thank you very much @romainwallon! 👍 

Since that worked, I've created https://github.com/kahypar/kahypar/pull/100 and cleaned up the fix a bit. Could you please try that code again before we merge this into master?

**Comment by romainwallon** [comment]
Hi @SebastianSchlag,

I tried to compile the code, and I had some problems when compiling, because the `-march=native` was still added to the compilation flags.
I found two ways for fixing it, that are described below.

The variable `CMAKE_APPLE_SILICON_PROCESSOR` is not set by default, and it has to be set *manually* as an environment variable or in the command line of `cmake` to the appropriate value (for M1 Macs, to `arm64`).
I did not know that when compiling the first time, so if you want to keep using this variable, maybe you could had this information to the description on how to build KaHyPar?

Another option would be to use `CMAKE_HOST_SYSTEM_PROCESSOR` instead in the `CMakeList.txt`, which would be equal to `arm64` on an M1 Mac, without having to set it.
However, I think it may be too generic to work has intended : it would probably be equal to `arm64` on any  computer equipped with an ARM processor, including computers that are not running macOS.
Thus, it is needed in this case to add a test to make sure that the system is indeed macOS.
To summarize this latter option, the idea would be to check that the processor is an M1 as follows:

```cmake
if(CMAKE_SYSTEM_NAME MATCHES "Darwin" AND CMAKE_HOST_SYSTEM_PROCESSOR MATCHES "arm64")
```

**Comment by SebastianSchlag** [comment]
Thanks @romainwallon  for testing, for the detailed explanation, and for the proposed fixes. I though that `CMAKE_APPLE_SILICON_PROCESSOR` would be set automatically.

I think using the second option (i.e., `if(CMAKE_SYSTEM_NAME MATCHES "Darwin" AND CMAKE_HOST_SYSTEM_PROCESSOR MATCHES "arm64")`) is probably best. I've updated the PR accordingly.

**Comment by jeshicawang** [comment]
Worked great for me! Thanks for looking into this @SebastianSchlag and @romainwallon.

**Comment by romainwallon** [comment]
It works fine for me too.

**Comment by SebastianSchlag** [comment]
Hey @limjiyoon,

Given your output, it looks like your python/miniconda installation is based on `x86_64`. Is this the case?
Could you try using the [Miniconda3 macOS Apple M1 ARM 64-bit bash](https://docs.conda.io/en/master/miniconda.html) version?

**Comment by limjiyoon** [comment]
Oh. Yes. It works!. Thanks for your help

**Comment by SebastianSchlag** [comment]
Thanks for confirming!
